File: ios/LocVault/Views/ConsentView.swift
import CoreLocation
import SwiftUI

struct ConsentView: View {
    @EnvironmentObject private var appState: AppState
    @State private var showPermissionInfo = false

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    statusCard
                    disclosureCard
                    actionButtons
                    auditLog
                }
                .padding()
            }
            .navigationTitle("Consent")
            .refreshable {
                await appState.refreshSession()
            }
            .sheet(isPresented: $showPermissionInfo) {
                PermissionInfoSheet {
                    showPermissionInfo = false
                    Task { await appState.grantConsent() }
                }
            }
        }
    }

    private var statusCard: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("Tracking status")
                    .font(.headline)
                Spacer()
                StatusBadge(active: appState.consent?.granted == true)
            }

            if let user = appState.user {
                Text("Signed in as \(user.name)")
                    .foregroundStyle(.secondary)
            }

            if appState.locationTracker.isTracking {
                Label("Reporting every 15 minutes", systemImage: "location.fill")
                    .font(.subheadline)
                    .foregroundStyle(.green)
            }

            if let lastReport = appState.locationTracker.lastReportAt {
                Text("Last report: \(lastReport.formatted(date: .abbreviated, time: .standard))")
                    .font(.footnote)
                    .foregroundStyle(.secondary)
            }

            if let error = appState.locationTracker.lastError {
                Text(error)
                    .font(.footnote)
                    .foregroundStyle(.red)
            }

            locationAuthHint
        }
        .padding()
        .background(Color(.secondarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    @ViewBuilder
    private var locationAuthHint: some View {
        switch appState.locationTracker.authorizationStatus {
        case .notDetermined:
            Text("Location permission will be requested when you grant consent.")
                .font(.footnote)
                .foregroundStyle(.secondary)
        case .authorizedWhenInUse:
            Text("Tip: Allow “Always” location access in Settings for background 15-minute reports.")
                .font(.footnote)
                .foregroundStyle(.orange)
        case .denied, .restricted:
            Text("Location access denied. Open Settings to enable location for Ground to Growth Connect.")
                .font(.footnote)
                .foregroundStyle(.red)
        default:
            EmptyView()
        }
    }

    private var disclosureCard: some View {
        VStack(alignment: .leading, spacing: 8) {
            if let disclosure = appState.disclosure {
                Text("Disclosure · v\(disclosure.version)")
                    .font(.headline)
                ScrollView {
                    Text(disclosure.text)
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
                .frame(maxHeight: 280)
            } else {
                ProgressView("Loading disclosure…")
            }
        }
        .padding()
        .background(Color(.secondarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    private var actionButtons: some View {
        VStack(spacing: 12) {
            Button {
                if appState.locationTracker.authorizationStatus == .notDetermined {
                    showPermissionInfo = true
                } else {
                    Task { await appState.grantConsent() }
                }
            } label: {
                Text("Grant consent")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.borderedProminent)
            .disabled(appState.consent?.granted == true || appState.isLoading)

            Button(role: .destructive) {
                Task { await appState.revokeConsent() }
            } label: {
                Text("Revoke consent")
                    .frame(maxWidth: .infinity)
            }
            .buttonStyle(.bordered)
            .disabled(appState.consent?.granted != true || appState.isLoading)
        }
    }

    @ViewBuilder
    private var auditLog: some View {
        if !appState.consentHistory.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                Text("Consent audit log")
                    .font(.headline)
                ForEach(appState.consentHistory) { record in
                    HStack {
                        Text(record.granted ? "Granted" : "Revoked")
                            .fontWeight(.medium)
                        Spacer()
                        Text(record.createdAt.prefix(19).replacingOccurrences(of: "T", with: " "))
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                    .padding(.vertical, 4)
                    Divider()
                }
            }
            .padding()
            .background(Color(.secondarySystemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 12))
        }
    }
}

struct PermissionInfoSheet: View {
    let onContinue: () -> Void
    @Environment(\.dismiss) private var dismiss

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 20) {
                    Image(systemName: "location.circle.fill")
                        .font(.system(size: 52))
                        .foregroundStyle(.tint)
                        .frame(maxWidth: .infinity)
                        .padding(.top, 8)

                    Text("iOS will ask you twice")
                        .font(.title2.bold())
                        .frame(maxWidth: .infinity, alignment: .center)

                    Text("To send location reports every 15 minutes — even in the background — iOS requires two steps. This is normal.")
                        .foregroundStyle(.secondary)

                    VStack(alignment: .leading, spacing: 16) {
                        stepRow(number: "1", title: "First prompt",
                                detail: "Tap “Allow While Using App.” (There is no “Always” option on the first prompt — Apple doesn’t allow it.)")
                        stepRow(number: "2", title: "Second prompt",
                                detail: "A moment later, tap “Change to Always Allow” so reports keep working in the background.")
                    }
                    .padding()
                    .background(Color(.secondarySystemBackground))
                    .clipShape(RoundedRectangle(cornerRadius: 12))

                    Text("You can change this anytime in Settings, and revoke consent to stop tracking immediately.")
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                }
                .padding()
            }
            .safeAreaInset(edge: .bottom) {
                VStack(spacing: 8) {
                    Button(action: onContinue) {
                        Text("Continue")
                            .frame(maxWidth: .infinity)
                    }
                    .buttonStyle(.borderedProminent)

                    Button("Not now") { dismiss() }
                        .buttonStyle(.borderless)
                }
                .padding()
                .background(.bar)
            }
            .navigationTitle("Location access")
            .navigationBarTitleDisplayMode(.inline)
        }
    }

    private func stepRow(number: String, title: String, detail: String) -> some View {
        HStack(alignment: .top, spacing: 12) {
            Text(number)
                .font(.headline)
                .foregroundStyle(.white)
                .frame(width: 28, height: 28)
                .background(Circle().fill(.tint))
            VStack(alignment: .leading, spacing: 2) {
                Text(title).font(.headline)
                Text(detail).font(.subheadline).foregroundStyle(.secondary)
            }
        }
    }
}

struct StatusBadge: View {
    let active: Bool

    var body: some View {
        Text(active ? "Active" : "Off")
            .font(.caption.bold())
            .padding(.horizontal, 10)
            .padding(.vertical, 4)
            .background(active ? Color.green.opacity(0.2) : Color.red.opacity(0.2))
            .foregroundStyle(active ? .green : .red)
            .clipShape(Capsule())
    }
}

File: mobile/lib/views/consent_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../app_state.dart';

class ConsentView extends StatelessWidget {
  const ConsentView({super.key});

  @override
  Widget build(BuildContext context) {
    return Consumer<AppState>(
      builder: (context, app, _) {
        return Scaffold(
          appBar: AppBar(title: const Text('Consent')),
          body: RefreshIndicator(
            onRefresh: app.refreshSession,
            child: ListView(
              padding: const EdgeInsets.all(16),
              children: [
                _StatusCard(app: app),
                const SizedBox(height: 16),
                _DisclosureCard(app: app),
                const SizedBox(height: 16),
                _ActionButtons(app: app),
                if (app.consentHistory.isNotEmpty) ...[
                  const SizedBox(height: 16),
                  _AuditLog(app: app),
                ],
              ],
            ),
          ),
        );
      },
    );
  }
}

class _Card extends StatelessWidget {
  final Widget child;
  const _Card({required this.child});

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: Theme.of(context).colorScheme.surfaceContainerHighest,
        borderRadius: BorderRadius.circular(12),
      ),
      child: child,
    );
  }
}

class _StatusCard extends StatelessWidget {
  final AppState app;
  const _StatusCard({required this.app});

  @override
  Widget build(BuildContext context) {
    final active = app.consent?.granted == true;
    return _Card(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              const Text('Tracking status', style: TextStyle(fontWeight: FontWeight.bold)),
              Container(
                padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 4),
                decoration: BoxDecoration(
                  color: (active ? Colors.green : Colors.red).withValues(alpha: 0.2),
                  borderRadius: BorderRadius.circular(20),
                ),
                child: Text(
                  active ? 'Active' : 'Off',
                  style: TextStyle(
                    color: active ? Colors.green : Colors.red,
                    fontWeight: FontWeight.bold,
                    fontSize: 12,
                  ),
                ),
              ),
            ],
          ),
          if (app.user != null) ...[
            const SizedBox(height: 8),
            Text('Signed in as ${app.user!.name}', style: const TextStyle(color: Colors.grey)),
          ],
        ],
      ),
    );
  }
}

class _DisclosureCard extends StatelessWidget {
  final AppState app;
  const _DisclosureCard({required this.app});

  @override
  Widget build(BuildContext context) {
    return _Card(
      child: app.disclosure == null
          ? const Center(child: Padding(padding: EdgeInsets.all(8), child: CircularProgressIndicator()))
          : Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text('Disclosure · v${app.disclosure!.version}',
                    style: const TextStyle(fontWeight: FontWeight.bold)),
                const SizedBox(height: 8),
                ConstrainedBox(
                  constraints: const BoxConstraints(maxHeight: 280),
                  child: SingleChildScrollView(
                    child: Text(app.disclosure!.text,
                        style: const TextStyle(color: Colors.grey, fontSize: 12)),
                  ),
                ),
              ],
            ),
    );
  }
}

class _ActionButtons extends StatelessWidget {
  final AppState app;
  const _ActionButtons({required this.app});

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        FilledButton(
          onPressed: (app.consent?.granted == true || app.isLoading) ? null : app.grantConsent,
          child: const SizedBox(width: double.infinity, child: Text('Grant consent', textAlign: TextAlign.center)),
        ),
        const SizedBox(height: 12),
        OutlinedButton(
          onPressed: (app.consent?.granted != true || app.isLoading) ? null : app.revokeConsent,
          style: OutlinedButton.styleFrom(foregroundColor: Colors.red),
          child: const SizedBox(width: double.infinity, child: Text('Revoke consent', textAlign: TextAlign.center)),
        ),
      ],
    );
  }
}

class _AuditLog extends StatelessWidget {
  final AppState app;
  const _AuditLog({required this.app});

  @override
  Widget build(BuildContext context) {
    return _Card(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          const Text('Consent audit log', style: TextStyle(fontWeight: FontWeight.bold)),
          const SizedBox(height: 8),
          for (final record in app.consentHistory) ...[
            Padding(
              padding: const EdgeInsets.symmetric(vertical: 4),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: [
                  Text(record.granted ? 'Granted' : 'Revoked',
                      style: const TextStyle(fontWeight: FontWeight.w500)),
                  Text(
                    record.createdAt.length >= 19
                        ? record.createdAt.substring(0, 19).replaceAll('T', ' ')
                        : record.createdAt,
                    style: const TextStyle(color: Colors.grey, fontSize: 12),
                  ),
                ],
              ),
            ),
            const Divider(),
          ],
        ],
      ),
    );
  }
}
